source/extensions/io_socket/user_space/io_socket_handle_impl.cc

```diff
@@ -0,0 +1,330 @@
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // Closed peer.
+  if (!peer_handle_) {
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // Error: write after close.
+  if (peer_handle_->isPeerShutDownWrite()) {
+    // TODO(lambdai): `EPIPE` or `ENOTCONN`.
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // The peer is valid but temporary not accepts new data. Likely due to flow control.
+  if (!peer_handle_->isWritable()) {
+    return {0, Api::IoErrorPtr(Network::IoSocketError::getIoSocketEagainInstance(),
+                               Network::IoSocketError::deleteIoError)};
+  }
+
+  auto* const dest_buffer = peer_handle_->getWriteBuffer();
+  // Write along with iteration. Buffer guarantee the fragment is always append-able.
+  uint64_t bytes_written = 0;
+  for (uint64_t i = 0; i < num_slice && !dest_buffer->highWatermarkTriggered(); i++) {
+    if (slices[i].mem_ != nullptr && slices[i].len_ != 0) {
+      dest_buffer->add(slices[i].mem_, slices[i].len_);
+      bytes_written += slices[i].len_;
+    }
+  }
+  peer_handle_->setNewDataAvailable();
+  ENVOY_LOG(trace, "socket {} writev {} bytes", static_cast<void*>(this), bytes_written);
+  return {bytes_written, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+}
+
+Api::IoCallUint64Result IoSocketHandleImpl::write(Buffer::Instance& buffer) {
+  // Empty input is allowed even though the peer is shutdown.
+  if (buffer.length() == 0) {
+    return Api::ioCallUint64ResultNoError();
+  }
+  if (!isOpen()) {
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // Closed peer.
+  if (!peer_handle_) {
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // Error: write after close.
+  if (peer_handle_->isPeerShutDownWrite()) {
+    // TODO(lambdai): `EPIPE` or `ENOTCONN`.
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // The peer is valid but temporary not accepts new data. Likely due to flow control.
+  if (!peer_handle_->isWritable()) {
+    return {0, Api::IoErrorPtr(Network::IoSocketError::getIoSocketEagainInstance(),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  uint64_t total_bytes_to_write = 0;
+  const uint64_t max_bytes_to_write = buffer.length();
+  while (peer_handle_->isWritable()) {
+    const auto& front_slice = buffer.frontSlice();
+    if (front_slice.len_ == 0) {
+      break;
```

Comment:
This doesn't seem correct, the presence of a size 0 slice in the buffer does not indicate that the buffer is empty.

Consider instead doing a single move with a size computed as a function of high watermark, see computation in the comment I added to readv.

Reply:
1) Let's take the readv implementation offline. 
2) frontSlice guaratee returns the non-empty one.
```
   * @return RawSlice the first non-empty slice in the buffer, or {nullptr, 0} if the buffer
   * is empty.
   * /
virtual RawSlice frontSlice() const PURE;
```

Reply:
Fixed: following the at least 16K + fill to highwatermark + 128K at most.